### ISSUE TITLE
Compiling gmic takes too long between producing output and uses too much memory

### ISSUE THREAD

**Issue by ryandesign** [dtschump/gmic#325]
I feel certain that this issue must have been reported to you by now, but I could not immediately find it here. Apologies if this is a duplicate.

[gmic is available in MacPorts](https://ports.macports.org/port/gmic). MacPorts has an automated build system which I maintain, and we have for years had problems getting that automated build system to successfully and reliably build gmic. The problems stem from what I perceive to be deficiencies in the way the gmic source code is organized and how the build system works.

The gmic build system compiles a single 15,000-line source code file four times (into the gmic executable, the gmic_gimp_gtk gimp plugin, the libgmic.a static library, and the libgmic.dylib or libgmic.so dynamic or shared library). Each of these four compilations can take nearly 4GB of memory and, on the MacPorts build system's 2009 Apple Xserve, can take 15 minutes. 8GB is the amount of RAM I typically allocate to the MacPorts build VMs (we have separate build VMs for over a dozen different macOS versions) because the build VMs have 8 CPU cores and our general assumption is that 1GB RAM should be enough for one compile process. There are certainly exceptions, and gmic currently appears to be one. When compiling gmic in parallel, even if only two jobs are happening simultaneously, this exceeds 8GB RAM usage (for compilation plus what the OS needs) therefore swapping occurs, which slows down the build and in some cases (presumably due to OS bugs in older versions of macOS) even causes the build VM to crash and need to be rebooted.

Our build system is designed with the assumption that a build will produce output from time to time, such as when each file is being compiled. A build that does not produce output for a long period of time is assumed to have failed in some way and is terminated. We used to set that period of time to 20 minutes, which caused many gmic builds to get terminated. Even after increasing the time period to 60 minutes we still see gmic builds terminating due to no build output.

For MacPorts I will recommend that we work around the problem by disabling parallel building for gmic.

I am not familiar with gmic at all so perhaps there is a reason why this is not possible but I recommend the gmic code be restructured into smaller units that can be compiled independently. Each separate file being compiled will then need less memory, and as you modify code the parts that have not been modified won't need to be recompiled. It should also be possible to reuse code: for example, it should be possible to compile the bulk of your code a single time and then link the resulting object files into a shared library and a static library. For the executable and the gimp plugin, it should be possible to link them with the shared or static library.

This was previously reported at https://trac.macports.org/ticket/54370 and https://github.com/dtschump/gmic-community/issues/85.

**Comment by dtschump** [comment]
Thanks for your report.
Here are my thoughts on this;

- I don't believe the code structure is in cause, having long files generally does not bother a compiler too much.
- What does bother the compiler **a lot** is the heavy use of templates, particularly during the optimization phase.
- You may try compiling without the optimization flags, and you'll probably see the difference. I'm interested by your feedback on this.
- The different G'MIC interfaces (plug-in, cli, etc.) are using different configurations because they are not using the same set of linked libraries (e.g. the plug-in don't need libjpeg or libtiff, while the cli tool does). So, I don't think it's really possible to have a single compiled unit that can be re-used for all interfaces, except if we split the source code in **a lot** of different units just to handle this issue.
- All in all, I would rather blame the compiler, rather than the G'MIC source code. Using `g++` on Linux does not cause this problem (it did few years ago, but it's mainly fixed now). A compiler that forces the developer to write his source code in a certain (constrained) way to eventually not eat all the memory during the compilation is definitely not a good compiler.
- Restructuring the G'MIC source code to split it into several units would take an infinite time, without any guarantees that your flawed compiler finally works better with the restructured version (as I said, I believe the problem comes from the use of templates and the compiler optimizer, not the length of the files). I hope you will understand that this is not something we have time to do for such an uncertain result.

So right now, I don't have any solutions, sorry.

**Comment by cooljeanius** [comment]
Would it help to get a time profile so we can figure out which compilation phase specifically is the one that's taking all the time?

**Comment by dtschump** [comment]
If possible yes, but I'm 99% sure this is due to the optimization phase.
There are a lot of template instanciation in the G'MIC code, and I guess this is one cause of the long optimization time.
Compiling G'MIC without optimization enabled is fast and does not eat a lot of memory.

**Comment by ryandesign** [comment]
> * Using `g++` on Linux does not cause this problem (it did few years ago, but it's mainly fixed now).

Thanks for this; it appears to be the same with Clang. I first noticed the problem in 2017 but since then Clang has improved and no longer has this problem.

I don't have memory consumption data available for all Xcode / Clang versions but I do have build times for gmic 3.0.2 on each of our build machines. With macOS 10.13.6 / Xcode 9.4.1 / Clang 902.0.39.2 and later, we saw reasonable build times (33-51 minutes), but with macOS 10.12.6 / Xcode 9.2 / Clang 900.0.39.2 and earlier builds took 2-3 times longer or were canceled after not producing any output for over an hour. In some cases this was due to swapping after running out of memory, but I also tried giving some build machines more RAM e.g. 15 GB RAM in one machine, the Clang process took 12.5 GB at its peak, swap usage was near 0, but the build was still canceled due to no output for over an hour.

In MacPorts we've addressed the problem by only using Xcode's Clang if it's 902 or greater; if an earlier Xcode Clang is found, a newer MacPorts Clang is used instead. That is a sufficient workaround as far as I'm concerned.

I assumed that the build enabled optimization because it was beneficial, so since I already had the above solution I didn't attempt any builds with lesser optimization levels.
